lib/ParseSIL/ParseSIL.cpp

```diff
@@ -3084,6 +3080,103 @@ bool SILParser::parseSILInstruction(SILBuilder &B) {
+    // Parse an index set, prefaced with the given label.
+    auto parseIndexSet = [&](StringRef label, SmallVectorImpl<unsigned> &indices,
+                             const Diagnostic &parseIndexDiag) -> bool {
+      // Parse `[<label> <integer_literal>...]`.
+      if (P.parseToken(tok::l_square, diag::sil_autodiff_expected_lsquare,
+                       "index list") ||
+          P.parseSpecificIdentifier(
+              label, diag::sil_autodiff_expected_index_list_label, label))
+        return true;
+      while (P.Tok.is(tok::integer_literal)) {
+        unsigned index;
+        if (P.parseUnsignedInteger(index, lastLoc, parseIndexDiag))
+          return true;
+        indices.push_back(index);
+      }
+      if (P.parseToken(tok::r_square, diag::sil_autodiff_expected_rsquare,
+                       "index list"))
+        return true;
+      return false;
+    };
+    // Parse parameter and result indices.
+    SmallVector<unsigned, 8> parameterIndices;
+    SmallVector<unsigned, 8> resultIndices;
+    // Parse parameter and result indices.
+    if (parseIndexSet("parameters", parameterIndices,
+                      diag::sil_autodiff_expected_parameter_index))
+      return true;
+    if (parseIndexSet("results", resultIndices,
+                      diag::sil_autodiff_expected_result_index))
+      return true;
+    // Parse a trailing 'where' clause (optional).
+    // This represents witness generic signature requirements.
+    GenericSignature *witnessGenSig = nullptr;
+    SourceLoc whereLoc;
+    SmallVector<RequirementRepr, 4> witnessRequirementReprs;
+    if (P.Tok.is(tok::l_square) && P.peekToken().is(tok::kw_where)) {
+      P.consumeToken(tok::l_square);
+      bool firstTypeInComplete;
+      P.parseGenericWhereClause(whereLoc, witnessRequirementReprs,
+                                firstTypeInComplete,
+                                /*AllowLayoutConstraints*/ false);
+      if (P.parseToken(tok::r_square, diag::sil_autodiff_expected_rsquare,
+                       "where clause"))
+        return true;
+    }
+    // Parse original function name.
+    SILFunction *originalFunction;
+    if (parseSILFunctionRef(InstLoc, originalFunction))
+      return true;
+    // Resolve witness requirements.
+    if (!witnessRequirementReprs.empty()) {
+      // Add original function generic parameters to a local scope so that they
+      // can be found during type resolution for `RequirementRepr`s.
+      Scope S(&P, ScopeKind::Generics);
+      auto *origGenEnv = originalFunction->getGenericEnvironment();
+      if (origGenEnv)
+        for (auto pd : origGenEnv->getGenericParams())
+          P.addToScope(pd->getDecl());
+      SmallVector<Requirement, 4> requirements;
+      auto *whereClause = TrailingWhereClause::create(
+          originalFunction->getModule().getASTContext(), whereLoc,
+          witnessRequirementReprs);
+      convertRequirements(originalFunction, whereClause->getRequirements(),
```

Comment:
Todo: fix `[where ...]` clause parsing. It doesn't work yet because `RequirementRepr`s are resolved to `Requirement`s with error types:
```
(error_type
  (original_type=generic_type_param_type depth=0 index=0 decl=main.(file).T@swift/test/AutoDiff/differentiability_witness_function_inst.sil:24:43))
<τ_0_0>
```

This is related to scope-based lookup: the generic parameters in the where clause come from the original function, which is outside of the body scope of the `differentiability_witness_function` instruction's function, so scope-based type resolution fails. I tried some strategies from parsing SIL differentiability witnesses but they didn't work, need more investigation.

Reply:
Fixed witness generic signature parsing by changing the syntax.

Previously, resolving witness generic requirements based on the original function's type/generic environment was problematic. If the original `SILFunction` is a bodiless declaration, it will not have a `GenericEnvironment`, so `where` clause `RequirementRepr` resolution will fail. Using the generic environment from the original function type's generic signature (`function->getLoweredFunctionType()->getGenericSignature()->getGenericEnvironment()`) did not work either.

```
// Old syntax.
differentiability_witness_function ... [where T: Differentiable] @foo : $(T) -> T
```

Now, the witness generic signature is parsed separately as a new generic environment with its own generic parameters and requirements:

```
// New syntax.
differentiability_witness_function ... <T where T: Differentiable> @foo : $(T) -> T
```

Parsing validation checks that the parsing witness generic signature is compatible with the original function generic signature (e.g. generic parameters are the same), then merges requirements from both to form the final witness generic signature.